In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
# Imports
import torch
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from tqdm import tqdm
from sklearn.metrics import accuracy_score, f1_score, classification_report
import google.generativeai as genai
import os
import time
import google.api_core.exceptions as gexc
import random as random
from requests.exceptions import HTTPError

import warnings
warnings.filterwarnings("ignore")

## Model selection & practical constraints

We selected Gemini-2.0-Flash for our experiments based on several practical considerations:

- **Free API access**: No cost barriers for limited exploration
- **Competitive reasoning abilities**: Comparable performance to GPT-4o on analytical tasks
- **Responsive latency**: Fast enough for batch processing within our timeframe

Our decision was significantly influenced by the API's usage limitations:
- Daily and weekly token quota restrictions
- Request frequency caps
- Per-request token limits

To work within these constraints while maintaining experimental validity, we:
- Sampled 500 comments from the dataset (large enough for statistical significance while respecting our token budget)
- Designed concise prompts (50-70 tokens each) to maximize the number of possible API calls
- Implemented a resilient request strategy that could navigate around rate limiting

In [ ]:
# Device & random seed
# genai.configure(api_key="AIzaSyAzQl2VLAg93xQTdma66Moyclb1uNoZ8Wg")
genai.configure(api_key="AIzaSyBVfiYq0mRQ0apDmMGq22rMsmqqF3v_uO8")

model = genai.GenerativeModel("gemini-2.0-flash")


DEVICE = "cuda" if torch.cuda.is_available() else "cpu"

SEED = 42
TRAIN_SIZE = 100
TEST_SIZE = 500

In [ ]:
# Load dataset
DATA_PATH = "/content/drive/MyDrive/Project/Finance_50K_Cleaned_Labels.csv"
df = (pd.read_csv(DATA_PATH, usecols=["preprocessed_text", "label", "clean_label"]).rename(columns={"preprocessed_text": "text", "label": "topic", "clean_label"  : "label"}))

In [ ]:
# sampling

df_train = df.sample(TRAIN_SIZE, random_state=SEED).reset_index(drop=True)            # few-shot pool
df_rest = df.drop(df_train.index).reset_index(drop=True)
df_test = df_rest.sample(TEST_SIZE, random_state=SEED+1).reset_index(drop=True)

## Prompt design strategy: optimizing model response qality

To achieve optimal performance from Gemini-2.0-Flash's limitations, we developed a specialized prompt format with a strategic arithmetic component that:

* Guides the model to provide concise, accurate responses
* Reduces policy-based rejections on sensitive content
* Improves classification consistency at low token overhead prompt structure

**All prompts follow this consistent pattern:**

Common structure for all three settings

**[TASK]  + [TOPIC]  + [COMMENT TEXT]  -> Return exactly ONE of {positive, neutral, negative}.**

**Zero-shot:** template only; relies completely on pre-training.

**One-shot:** template + one labeled example (sampled from the training split, never from the test set) to determine the output format.

**Few-shot:** template + three labelled examples randomly drawn from the training pool. This gives the model additional guidance while keeping the prompt short and within the context window.

**Why three examples?**

Every additional demo adds ~80 tokens. With 500 questions and a daily budget of tokens, three demos stay within quota.
Early tests showed that using more than three examples didn’t improve accuracy, so extra tokens would be wasted.
All three versions use the same zero-shot template, making it easy to change the number of examples in code.

In [ ]:
class PromptBuilder:
    def __init__(self, df_pool: pd.DataFrame, k_few: int = 3):
        self.pool = df_pool.reset_index(drop=True)
        self.one  = self.pool.sample(1, random_state=SEED).reset_index(drop=True)
        self.few  = self.pool.sample(k_few, random_state=SEED+1).reset_index(drop=True)

    def _fmt(self, txt: str) -> str:
        return txt.replace("\n", " ")[:160]

    # zero-shot
    def zero(self, x: str, t: str) -> str:
        x, t = self._fmt(x), self._fmt(t)
        return (
            "Task: learn f: C → S,  S = {positive, negative, neutral}\n"
            f'Topic : "{t}"\n'
            f'Input x: "{x}"\n'
            "Output y ∈ S (ONE WORD exactly):"
        )

    # one-shot
    def one_shot(self, x: str, t: str) -> str:
        ex = self.one.iloc[0]
        ex_txt, ex_lab, ex_top = self._fmt(ex.text), ex.label, self._fmt(ex.topic)
        x, t = self._fmt(x), self._fmt(t)
        return (
            "Task: learn f: C → S,  S = {positive, negative, neutral}\n"
            f'Example topic₁: "{ex_top}"\n'
            f'Example x₁   : "{ex_txt}" → y₁ = {ex_lab}\n'
            f'Topic        : "{t}"\n'
            f'Input   x    : "{x}"\n'
            "Output  y ∈ S (ONE WORD exactly):"
        )

    # few-shot
    def few_shot(self, x: str, t: str) -> str:
        shots = [
            f'x{i}: "{self._fmt(r.text)}"  (topic="{self._fmt(r.topic)}") → y{i} = {r.label}'
            for i, r in enumerate(self.few.itertuples(), start=1)
        ]
        x, t = self._fmt(x), self._fmt(t)
        return (
            "Task: learn f: C → S,  S = {positive, negative, neutral}\n"
            + "\n".join(shots) + "\n"
            f'Topic : "{t}"\n'
            f'Input x: "{x}"\n'
            "Output y ∈ S (ONE WORD exactly):"
        )


## API request implementation
Our *run_gemini* helper function effectively manages API communication with two key techniques:

- Random timing variations between requests (200-800ms) to prevent pattern detection
- Smart retry strategy that increases wait times exponentially after three or more failed attempts

These techniques maintained a request success rate above 95% throughout our 500 sample experiment, and preventing permanent API blocks.

In [ ]:
SLEEP_SEC = 3
JITTER = 1.5

def run_gemini(prompt):
    max_retries = 7
    base_wait = 5
    for attempt in range(max_retries):
        try:
            resp = model.generate_content(prompt)
            wait_time = SLEEP_SEC + random.uniform(0, JITTER)
            time.sleep(wait_time)
            return extract_label(resp.text)
        except Exception as e:
            wait_time = base_wait * (2 ** attempt) + random.uniform(1, 3)
            print(f"Error: {type(e).__name__}. Retrying in {wait_time:.1f}s (attempt {attempt+1}/{max_retries})")
            time.sleep(wait_time)
    print("All retries failed")
    return "error"

def extract_label(text: str) -> str:
    tok = text.lower().strip().split()[0].strip(".,")
    return tok if tok in {"positive","negative","neutral"} else "error"

In [ ]:
def evaluate(builder_fn, test_df: pd.DataFrame, k: int | None = None, seed: int = 42) -> pd.DataFrame:
    label, pred = [], []
    if k is None or k > len(test_df):
        subset = test_df
    else:
        subset = test_df.sample(k, random_state=seed)

    for row in tqdm(subset.itertuples(), total=len(subset)):
        label.append(row.label.lower())
        prompt = builder_fn(row.text, row.topic)
        pred.append(run_gemini(prompt))
    return pd.DataFrame({"label": label, "pred": pred})

In [ ]:
builder = PromptBuilder(df_train, k_few=3)

## Zero-shot prompt
*Template* – *topic* + original Reddit comment + “**Return one of {positive, neutral, negative}.**”  
We deliberately **did not** supply an example so the model must rely purely on its pre-training knowledge.

In [ ]:
df_zero = evaluate(builder.zero, df_test, k=TEST_SIZE)

100%|█████████▉| 499/500 [36:18<00:04,  4.34s/it]WARNING:tornado.access:429 POST /v1beta/models/gemini-2.0-flash:generateContent?%24alt=json%3Benum-encoding%3Dint (127.0.0.1) 380.84ms


Error: TooManyRequests. Retrying in 6.5s (attempt 1/7)


Error: TooManyRequests. Retrying in 11.8s (attempt 2/7)


Error: TooManyRequests. Retrying in 21.8s (attempt 3/7)


Error: TooManyRequests. Retrying in 43.0s (attempt 4/7)


Error: TooManyRequests. Retrying in 81.8s (attempt 5/7)


Error: TooManyRequests. Retrying in 162.9s (attempt 6/7)


Error: TooManyRequests. Retrying in 322.6s (attempt 7/7)


100%|██████████| 500/500 [47:11<00:00,  5.66s/it] 

All retries failed


In [ ]:
print(classification_report(df_zero.label, df_zero.pred))

df_zero.to_csv(f"sentiment_zero_shot_results_{time.strftime('%Y%m%d_%H%M')}.csv", index=False)
df_test.to_csv(f"sentiment_test_with_predictions_{time.strftime('%Y%m%d_%H%M')}.csv", index=False)

              precision    recall  f1-score   support

       error       0.00      0.00      0.00         0
    negative       0.65      0.90      0.76       185
     neutral       0.79      0.51      0.62       218
    positive       0.59      0.63      0.61        97

    accuracy                           0.68       500
   macro avg       0.51      0.51      0.50       500
weighted avg       0.70      0.68      0.67       500



*Observation* – High recall on the **negative** class (0.90) but poor recall on **positive** (0.63) and **neutral** (0.51) suggests the language model defaults to a cautious negative or neutral stance without guidance.

## One-shot prompt
We added **one example** (taken from the training split, unseen in test) because a single concrete example nudges the model toward the task goal while keeping token usage minimal.

In [ ]:
df_one  = evaluate(builder.one_shot,  df_test, k=TEST_SIZE)

100%|██████████| 500/500 [36:01<00:00,  4.32s/it]


In [ ]:
print(classification_report(df_one.label, df_one.pred))

df_one.to_csv(f"sentiment_one _shot_results_{time.strftime('%Y%m%d_%H%M')}.csv", index=False)
df_test.to_csv(f"sentiment_test_with_predictions_{time.strftime('%Y%m%d_%H%M')}.csv", index=False)

              precision    recall  f1-score   support

    negative       0.78      0.73      0.75       185
     neutral       0.64      0.80      0.71       218
    positive       0.72      0.40      0.52        97

    accuracy                           0.70       500
   macro avg       0.71      0.64      0.66       500
weighted avg       0.71      0.70      0.69       500



### Results

| metric | value | Δ vs 0-shot |
|--------|-------|------------|
| accuracy | **0.698** | +0.022 |
| macro-F1 | **0.661** | +0.166 |

*Interpretation* – The macro-F1 score went up by about 17 points. Adding only one example with a label helps Gemini understand all three labels better and pick “positive” much more correctly (precision goes from 0.59 to 0.72).


## Few-shot prompt (3 examples)

We used **three random examples** in the prompt. We couldn’t add more because each example adds about 80 tokens and we’d hit our request limit too fast.


In [ ]:
df_few  = evaluate(builder.few_shot,  df_test, k=TEST_SIZE)

100%|██████████| 500/500 [35:53<00:00,  4.31s/it]


In [ ]:
print(classification_report(df_few.label, df_few.pred))

df_few.to_csv(f"sentiment_few_shot_results_{time.strftime('%Y%m%d_%H%M')}.csv", index=False)
df_test.to_csv(f"sentiment_test_with_predictions_{time.strftime('%Y%m%d_%H%M')}.csv", index=False)

              precision    recall  f1-score   support

    negative       0.70      0.91      0.79       185
     neutral       0.84      0.69      0.76       218
    positive       0.82      0.66      0.73        97

    accuracy                           0.77       500
   macro avg       0.79      0.76      0.76       500
weighted avg       0.78      0.77      0.77       500



##### Results

| metric | value | Δ vs 1-shot |
|--------|-------|------------|
| accuracy | **0.768** | +0.070 |
| macro-F1 | **0.761** | +0.100 |

*Interpretation*   

* Accuracy reached 0.77, which is 9 points higher than zero-shot.
* Macro-F1 went up to 0.76, showing the model handles rare labels better.


In [ ]:
def score(df):
    acc = accuracy_score(df.label, df.pred)
    f1  = f1_score(df.label, df.pred, average="macro", zero_division=0)
    return acc, f1

for name, df_res in [("zero", df_zero), ("one", df_one), ("few", df_few)]:
    acc, f1 = score(df_res)
    print(f"{name:4s}: acc = {acc:.3f}    macro-F1 = {f1:.3f}")


zero: acc = 0.676    macro-F1 = 0.495
one : acc = 0.698    macro-F1 = 0.661
few : acc = 0.768    macro-F1 = 0.761


In [ ]:
load_test_df = pd.read_csv("sentiment_test_with_predictions_20250517_1428.csv")
load_zero_df = pd.read_csv("sentiment_zero_shot_results_20250517_1347.csv")
load_one_df = pd.read_csv("sentiment_one_shot_results_20250517_1428.csv")
load_few_df = pd.read_csv("sentiment_few_shot_results_20250517_1517.csv")

In [ ]:
def score(df):
    acc = accuracy_score(df.label, df.pred)
    f1  = f1_score(df.label, df.pred, average="macro", zero_division=0)
    return acc, f1

for name, df_res in [("zero", load_zero_df), ("one", load_one_df), ("few", load_few_df)]:
    acc, f1 = score(df_res)
    print(f"{name:4s}: acc = {acc:.3f}    macro-F1 = {f1:.3f}")


zero: acc = 0.676    macro-F1 = 0.495
one : acc = 0.698    macro-F1 = 0.661
few : acc = 0.768    macro-F1 = 0.761


## Summary of LLM sentiment runs

| setting | accuracy | macro-F1 |
|---------|----------|----------|
| zero-shot | 0.676 | 0.495 |
| one-shot  | 0.698 | 0.661 |
| few-shot  | 0.768 | 0.761 |

Adding examples steadily improves performance, with the biggest boost coming from zero-shot to one-shot.

Recall for the positive class is still low (0.66). Adding more positive examples could help.

Using a model with a larger context window (like Gemini-Pro) might let us try 5-shot prompts without hitting limits.
